src/toil_lib/test/test_programs.py

```diff
@@ -1,14 +1,70 @@
+def test_docker_clean(tmpdir):
+    """
+    Run the test container that creates a file in the work dir, and sleeps for 5 minutes.  Ensure
+    that the calling job gets SIGKILLed after a minute, leaving behind the spooky/ghost/zombie
+    container. Ensure that the container is killed on batch system shutdown (through the defer
+    mechanism).
+
+    This inherently also tests docker_call
+
+    :param tmpdir: A temporary directory to work in
+    :returns: None
+    """
+    assert os.getuid() != 0, "Cannot test this if the user is root."
+    data_dir = os.path.join(str(tmpdir), 'data')
+    work_dir = os.path.join(str(tmpdir), 'working')
+    os.mkdir(data_dir)
+    os.mkdir(work_dir)
+    options = Job.Runner.getDefaultOptions(os.path.join(str(tmpdir), 'jobstore'))
+    options.logLevel = 'INFO'
+    options.workDir = work_dir
+    A = Job.wrapJobFn(_test_docker_clean_fn, data_dir)
+    try:
+        Job.Runner.startToil(A, options)
+    except FailedJobsException:
+        # The file created by spooky_container would remain in the directory, and since it was
+        # created inside the container, it would have had uid and gid == 0 (root) upon creation.
+        # If the defer mechanism worked, it should now be non-zero and we check for that.
+        file_stats = os.stat(os.path.join(data_dir, 'test.txt'))
+        assert file_stats.st_gid != 0
+        assert file_stats.st_uid != 0
+        # Ensure the container isn't still running.
+        with open(os.path.join(data_dir, 'container.name'), 'r') as file_handle:
+            container_name = file_handle.read()
+        container_prefix = '--'.join(container_name.split('--')[:2])
+        existing_containers = subprocess.check_output(['docker', 'ps', '-a', '--format',
+                                                       '"{{.Names}}"'])
+        assert container_prefix not in existing_containers
+
+
+def _test_docker_clean_fn(job, data_dir):
+    def kill_self():
+        test_file = os.path.join(data_dir, 'test.txt')
+        # This will kill the worker once we are sure the docker container started
+        while not os.path.exists(test_file):
+            _log.debug('Waiting on the file created by spooky_container.')
+            time.sleep(1)
+        # By the time we reach here, we are sure the container is running.  Write down the container
+        # prefix so we can check for it later.
+        with open(os.path.join(data_dir, 'container.name'), 'w') as file_handle:
+            file_handle.write(_get_container_name(job))
+        os.kill(os.getpid(), signal.SIGKILL)  # signal.SIGINT)
+        # TODO: FIX OCCASIONAL DEADLOCK WITH SIGINT (tested on single machine)
+    t = Thread(target=kill_self)
+    # Make it a daemon thread so that thread failure doesn't hang tests.
+    t.daemon = True
+    t.start()
+    docker_call(job, tool='quay.io/ucsc_cgl/spooky_test', work_dir=data_dir, detached=False,
+                rm=True, defer=docker_call.RM)
```

Comment:
If you have the time, can you think about whether testing `defer=STOP` with `rm=True` and `rm=False` can be done without adding a lot of test code?